### PR TITLE
On-Chain PCCS Audit Submission For Release (November 2024)

### PR DESCRIPTION
**Pull request by preston4896**
# Overview

New changes included in this PR that we are planning to roll out for the next release of both [Automata On-Chain PCCS](https://github.com/automata-network/automata-on-chain-pccs) and [Automata DCAP Attestation](https://github.com/automata-network/automata-dcap-attestation).

# Major change(s) included in this PR:

- Dynamic P256 Verifier configuration upon deployment. See relevant changes: 
[`ad672b9`](https://github.com/automata-network/automata-on-chain-pccs/commit/ad672b9206fc5c971757204d082b73d27921e1d2)
[`70c4032`](https://github.com/automata-network/automata-on-chain-pccs/commit/70c4032a80b590dd383392fcf820d071fbcb8586)

- Improved collateral data management, using the `IDaoAttestationResolver` interface. See relevant changes: 
[`a25d208`](https://github.com/automata-network/automata-on-chain-pccs/commit/a25d20843e819638b757db4724fec919cf35e501)
[`376b1e3`](https://github.com/automata-network/automata-on-chain-pccs/commit/376b1e39fd4e7d86e127fffb188b5f774acb3480)

- temporary caller restriction for reading collaterals from PCCS, see relevant changes:
[`8f73ebe`](https://github.com/automata-network/automata-on-chain-pccs/commit/8f73ebeaad0d6c2a3801f44ba5124091985f3c76)
[`a103fc1`](https://github.com/automata-network/automata-on-chain-pccs/commit/a103fc1423502971ff2be7fad0bfa8f2881f1439)
[`1a389e1`](https://github.com/automata-network/automata-on-chain-pccs/commit/1a389e11e26b9468475a0c15a54a90acaae7bc8a)
[`30301fa`](https://github.com/automata-network/automata-on-chain-pccs/commit/30301faa2976fc61ebc69ce8338544e73121ea2b)

# When auditing is in progress, expect the following:
- Upon finalization of this PR (Converted to Ready To Review status), there shall be no more major breaking changes introduced to this PR. (e.g. external API changes)
- Code optimizations may require minor changes on internal methods.
- Revision(s) that may be necessary based on findings from auditing. (May be exempted from the "no breaking changes" rule)

### CODE REVIEW

File: src/automata_pccs/AutomataPcsDao.sol [hunk 1]
```diff
@@ -1,41 +1,22 @@
...
 
-import {AutomataDaoBase} from "./shared/AutomataDaoBase.sol";
-import {PcsDao, AttestationRequest, X509CRLHelper} from "../bases/PcsDao.sol";
-
+import {PcsDao, X509CRLHelper, DaoBase} from "../bases/PcsDao.sol";
 import {Ownable} from "solady/auth/Ownable.sol";
 
+import {AutomataDaoBase} from "./shared/AutomataDaoBase.sol";
+
 contract AutomataPcsDao is AutomataDaoBase, PcsDao, Ownable {
```
[the remark this answers is not in the document]
Comment (Liao1) [reply src/automata_pccs/AutomataPcsDao.sol:9]: New deployment to testnet:
- re-run existing transactions

File: src/automata_pccs/shared/AutomataDaoStorage.sol [hunk 2]
```diff
@@ -25,25 +53,90 @@ contract AutomataDaoStorage is Ownable {
...
+        returns (bytes memory attData)
+    {
+        attData = _db[attestationId];
     }
 
-    function readPccs(bytes32 attId) external view returns (bytes memory attData) {
-        attData = _pccsData[attId];
+    /**
+     * @notice In AutomataDaoStorage, we will simply assign the key as the attestationid of the collateral
+     * @notice whereas the value (key + 1) will be the attestation id to the hash.
```
Comment (Liao1) [inline src/automata_pccs/shared/AutomataDaoStorage.sol:79]: add a prefix to the pre-image to differentiate collateral and hash attestations

File: src/bases/EnclaveIdentityDao.sol [hunk 3]
```diff
@@ -10,156 +10,144 @@ import {DaoBase} from "./DaoBase.sol";
...
      */
     function _validateQeIdentity(EnclaveIdentityJsonObj calldata enclaveIdentityObj) private view {
         // Get TCB Signing Cert
-        bytes32 tcbSigningAttestationId = Pcs.pcsCertAttestations(CA.SIGNING);
-        bytes memory signingDer = getAttestedData(tcbSigningAttestationId);
+        // bytes memory signingDer = _fetchDataFromResolver(Pcs.PCS_KEY(CA.SIGNING, false), false);
+        // TEMP: calling _fetchDataFromResolver() would make more sense semantically
+        // TEMP: but I am calling the resolver directly here
+        // TEMP: so _fetchDataFromResolver() can be overridden without breaking here...
+        bytes memory signingDer = resolver.readAttestation(resolver.collateralPointer(Pcs.PCS_KEY(CA.SIGNING, false)));
```
Comment (Liao1) [inline src/bases/EnclaveIdentityDao.sol:150]: define a default internal read()

File: src/bases/FmspcTcbDao.sol [hunk 4]
```diff
@@ -79,85 +71,65 @@ abstract contract FmspcTcbDao is DaoBase, SigVerifyBase {
...
+        virtual
         returns (bytes32 attestationId)
     {
-        attestationId = fmspcTcbInfoAttestations[keccak256(abi.encodePacked(uint8(tcbType), fmspc, uint32(version)))];
+        (attestationId,) = resolver.attest(key, reqData, hash);
     }
 
     /**
-     * @notice builds an EAS compliant attestation request
+     * @notice constructs the Identity.json attestation data
```
Comment (Liao1) [inline src/bases/FmspcTcbDao.sol:120]: TcbInfo.json

File: src/automata_pccs/AutomataPcsDao.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (preston4896) [reply src/automata_pccs/AutomataPcsDao.sol]: resolved [8da2565](https://github.com/automata-network/automata-on-chain-pccs/pull/9/commits/8da256593b50d9e84929c02f53949b11a1962232)

File: src/bases/EnclaveIdentityDao.sol (hunk 3 above)
Comment (preston4896) [reply src/bases/EnclaveIdentityDao.sol]: resolved [1579be8](https://github.com/automata-network/automata-on-chain-pccs/pull/9/commits/1579be8a79870b33f39a1ee800baeb7ce5a792b7)

File: src/bases/FmspcTcbDao.sol (hunk 4 above)
Comment (preston4896) [reply src/bases/FmspcTcbDao.sol]: resolved [83024c1](https://github.com/automata-network/automata-on-chain-pccs/pull/9/commits/83024c13e2a448f11f00796198308ed36465d829)

File: src/automata_pccs/shared/AutomataDaoStorage.sol (hunk 2 above)
Comment (preston4896) [reply src/automata_pccs/shared/AutomataDaoStorage.sol:79]: resolved [5d7dcbe](https://github.com/automata-network/automata-on-chain-pccs/pull/9/commits/5d7dcbe67437f94ff7d4603b727ca8e2764b65e6)

File: src/automata_pccs/shared/AutomataDaoStorage.sol [hunk 5]
```diff
@@ -1,20 +1,48 @@
...
+
+    function isAuthorizedCaller(address caller) external view returns (bool) {
+        return _authorized_readers[caller];
+    }
+
+    function setCallerAuthorization(address caller, bool authorized) external onlyOwner {
+        _authorized_readers[caller] = authorized;
+    }
+
+    function pauseCallerRestriction() external onlyOwner {
```
[the remark this answers is not in the document]
Comment (preston4896) [reply src/automata_pccs/shared/AutomataDaoStorage.sol]: resolved [3b11c53](https://github.com/automata-network/automata-on-chain-pccs/pull/9/commits/3b11c533575d4ea9b49d805bba27a3df40165801)

File: src/automata_pccs/shared/AutomataDaoStorage.sol [hunk 6]
```diff
@@ -1,20 +1,48 @@
...
-    mapping(bytes32 attId => bytes attData) _pccsData;
+/**
+ * @title Automata PCCS Dao Storage
+ * @notice This contract resolves and stores all collateral data internally
+ */
 
-    modifier onlyAuthorized(address dao) {
-        require(_authorized[dao], "Unauthorized caller");
+contract AutomataDaoStorage is AutomataTCBManager, IDaoAttestationResolver, Pausable, Ownable {
+    mapping(address => bool) _authorized_writers;
```
[the remark this answers is not in the document]
Comment (preston4896) [reply src/automata_pccs/shared/AutomataDaoStorage.sol:16]: Dismissing all name-convention issues

**Pull request merged**